File: Bitmask/0805.Split Array With Same Average/Solution.py
class Solution:
    def splitArraySameAverage(self, nums: List[int]) -> bool:
        n = len(nums)
        if n == 1:
            return False
        m, s = n // 2, sum(nums)
        for i in range(n):
            nums[i] = nums[i] * n - s
        vis = set()
        for i in range(1, 1 << m):
            ss = 0
            for j in range(m):
                if (1 << j) & i:
                    ss += nums[j]
            if ss == 0:
                return True
            vis.add(ss)
        for i in range(1, 1 << (n - m)):
            ss = 0
            for j in range(n - m):
                if (1 << j) & i:
                    ss += nums[j + m]
            if ss == 0 or (i != (1 << n - m) - 1 and -ss in vis):
                return True
        return False

File: Bitmask/0805.Split Array With Same Average/Solution.java
class Solution {
    public boolean splitArraySameAverage(int[] nums) {
        int n = nums.length, m = n / 2, s = 0;
        if (n == 1) {
            return false;
        }
        for (int v : nums) {
            s += v;
        }
        for (int i = 0; i < n; i++) {
            nums[i] = nums[i] * n - s;
        }
        Set<Integer> vis = new HashSet<>();
        for (int i = 1; i < (1 << m); i++) {
            int ss = 0;
            for (int j = 0; j < m; j++) {
                if (((1 << j) & i) != 0) {
                    ss += nums[j];
                }
            }
            if (ss == 0) {
                return true;
            }
            vis.add(ss);
        }
        for (int i = 1; i < (1 << (n - m)); i++) {
            int ss = 0;
            for (int j = 0; j < (n - m); j++) {
                if (((1 << j) & i) != 0) {
                    ss += nums[j + m];
                }
            }
            if (ss == 0 || (i != (1 << (n - m)) - 1 && vis.contains(-ss))) {
                return true;
            }
        }
        return false;
    }
}
